### PR TITLE
feat: add permitOnlyProjectScopedClusters flag

### LINKED ISSUE #10220: Permit only Project-scoped clusters to be a valid Destination
# Summary

It seems to be the case that project-scoped `Application`s can be installed onto clusters which are not a part of the same project.

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: "some-ns"
spec:
  destination:
    # This destination might not actually be a cluster which belongs to `foo-project`
    server: https://some-k8s-server/
    namespace: "some-ns"
  project: foo-project
```

# Motivation

So far this has not been a problem, but it would be nice to ensure that this situation cannot happen at all. We have a bunch of developers that are slinging `Applications` left and right, and we'd like to ensure that mistakes (or something more... sinister) cannot happen.

# Proposal

How do you think this should be implemented?

In `AppProjectSpec` we would add an extra attribute, `permitOnlyProjectScopedClusters` or something similar.

Then in `app_project_types.go`, we can extend the existing validation we have on `IsDestinationPermitted` to be something like this: 

```go
func (proj AppProject) IsDestinationPermitted(dst ApplicationDestination, projectClusters func(project string) ([]*Cluster, error)) (bool, error) {
        // The existing logic we have for isDestinationPermitted
	destinationMatched := proj.isDestinationPermitted(dst) 
	if destinationMatched && proj.Spec.PermitOnlyProjectScopedClusters {
		clusters, err := projectClusters(proj.Name)
		if err != nil {
			return false, err
		}

		// TODO: If no project clusters are found, return an error instead?
		for _, cluster := range clusters {
			if cluster.Server == dst.Server {
				return true, nil
			}
		}

		return false, nil
	}

	return destinationMatched, nil
}
```

`projectClusters` is a function which would wrap `ArgoDB.GetProjectClusters`, where we'd get all clusters for the given `AppProject` and then check if that a destination is indeed a part of the given `AppProject`.

### PR DESCRIPTION
**Pull request by blakepettersson**
This commit adds a new flag, `permitOnlyProjectScopedClusters`, which
prevents any application from syncing to clusters which are not a part
of the same project. Fixes #10220.

Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* ~[ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.~
* [x] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).

### CODE REVIEW

File: controller/appcontroller.go
Comment (blakepettersson) [inline controller/appcontroller.go:429]: I'm not a big fan of silencing errors, we should probably change the signature of `IterateHierarchy` to return an error itself

File: controller/appcontroller.go
Comment (blakepettersson) [inline controller/appcontroller.go:458]: Likewise, same as above

File: controller/state.go
Comment (blakepettersson) [inline controller/state.go:421]: I added `context.TODO()` as a placeholder, but should probably be something else

File: controller/sync.go
Comment (blakepettersson) [inline controller/sync.go:224]: I added context.TODO() as a placeholder, but should probably be something else

File: controller/state.go
Comment (blakepettersson) [inline controller/state.go:424]: Not sure if this block is the right way to do this; I suspect that this also needs tests

File: pkg/apis/application/v1alpha1/app_project_types.go
Comment (crenshaw-dev) [inline pkg/apis/application/v1alpha1/app_project_types.go:383]: Should this error be wrapped?

File: pkg/apis/application/v1alpha1/app_project_types.go
Comment (crenshaw-dev) [inline pkg/apis/application/v1alpha1/app_project_types.go:386]: I feel like simply returning "not permitted" is probably sufficient.

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]
Just a couple nitpicks. @alexmt would you have time to glance at this? I'm curious if you have any performance concerns.

File: pkg/apis/application/v1alpha1/app_project_types.go
Comment (blakepettersson) [reply pkg/apis/application/v1alpha1/app_project_types.go]: 👍

**Comment by crenshaw-dev** [comment]
@blakepettersson we discussed this in today's [security meeting](https://docs.google.com/document/d/1nl1adCRf4tD87Y01i3JcrLcHvtm1O5a7slqzNstct94/edit#).

The conclusions were:
1) This PR does solve a problem 😄 
2) Users might be surprised that this behavior isn't the default (so it might be helpful to add a line in the project-scoped cluster docs saying "this feature enables _adding_ clusters, but it does not restrict which clusters may be used. To do that you need to use the destinations field or the permitOnlyProjectScopedClusters flag"
3) @jessesuen wanted to ponder the problem a bit more and see if he could think of a solution that doesn't require a new flag - will wait for his conclusion before merging
4) performance might be a concern, so Jesse and/or @alexmt should take a look

Thanks again for the good work! Assuming we conclude that the change is necessary/performant, I wanna be sure to get this into 2.5, so I'll be watching carefully.

**Comment by blakepettersson** [comment]
Awesome! Looking forward to what you guys come up with! 😄 

> Users might be surprised that this behavior isn't the default 

I was one of them 😄

**Comment by crenshaw-dev** [comment]
Looks like something set you up for a bunch of merge conflicts. Can you resolve? It'll probably be easiest to resolve conflicts for the non-generated files and then just run codegen again.

**Comment by crenshaw-dev** [comment]
Discussed this in the Contributor office hours. @jessesuen and @alexmt are gonna chat about this and review.

**Comment by blakepettersson** [comment]
@jessesuen @alexmt @crenshaw-dev do you guys have any further thoughts on this PR? :smile:

**Review by crenshaw-dev — APPROVED** [review]
LGTM - will leave this open in case Jesse or Alex have any thoughts, but I'll merge it by the end of the week if not.

**Comment by blakepettersson** [comment]
Awesome, thanks a lot!! :hugs: :hugs:

**Pull request merged**
